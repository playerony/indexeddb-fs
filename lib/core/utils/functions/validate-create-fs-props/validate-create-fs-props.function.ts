import { validate } from 'jsonschema';

import { CreateFSProps } from '@core';

import { createFSPropsSchema } from './validate-create-fs-props.schema';

export function validateCreateFSProps(props: Required<CreateFSProps>): boolean {
  const { valid, errors } = validate(props, createFSPropsSchema);

  if (valid) {
    return valid;
  }

  const errorsAsString = JSON.stringify(errors);

  throw new Error(`Props passed to createFS function are invalid:\n${errorsAsString}`);
}
