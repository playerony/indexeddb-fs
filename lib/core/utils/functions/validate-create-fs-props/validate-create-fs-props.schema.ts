const stringRegExp = '^[a-zA-Z0-9_.-]*$';

export const createFSPropsSchema = {
  type: 'object',
  id: '/CreateFSPropsSchema',
  properties: {
    databaseName: {
      minLength: 4,
      maxLength: 50,
      type: 'string',
      pattern: stringRegExp,
    },
    objectStoreName: {
      minLength: 1,
      maxLength: 20,
      type: 'string',
      pattern: stringRegExp,
    },
    rootDirectoryName: {
      minLength: 1,
      maxLength: 20,
      type: 'string',
      pattern: stringRegExp,
    },
    databaseVersion: {
      minimum: 1,
      maximum: 100,
      type: 'integer',
    },
  },
  required: ['databaseName', 'databaseVersion', 'objectStoreName', 'rootDirectoryName'],
};
