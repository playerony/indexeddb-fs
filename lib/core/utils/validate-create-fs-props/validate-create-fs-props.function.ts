import { validate } from 'jsonschema';

import { CreateFsProps } from '@core';

import { createFsPropsSchema } from './validate-create-fs-props.schema';

export function validateCreateFSProps(props: Required<CreateFsProps>): boolean {
  const { valid, errors } = validate(props, createFsPropsSchema);

  if (valid) {
    return valid;
  }

  const errorsAsString = JSON.stringify(errors);

  throw new Error(`Props passed to createFS function are invalid:\n${errorsAsString}`);
}
