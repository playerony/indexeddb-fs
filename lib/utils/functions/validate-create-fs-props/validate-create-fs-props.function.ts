import { validate } from 'jsonschema';

import { CreateFsProps } from '@framework/create-fs.types';

import { createFsPropsSchema } from './validate-create-fs-props.schema';

export function validateCreateFSProps(props: Required<CreateFsProps>): boolean {
  const { valid, errors } = validate(props, createFsPropsSchema);

  if (valid) {
    return valid;
  }

  const errorsAsString = JSON.stringify(errors);

  throw new Error(`Props passed to createFS function are invalid:\n${errorsAsString}`);
}
