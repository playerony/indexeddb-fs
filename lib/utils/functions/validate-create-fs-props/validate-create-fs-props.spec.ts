import { toMatchSnapshot, functionImportTest } from '@utils';

import { createFsPropsSchema } from './validate-create-fs-props.schema';

import { validateCreateFSProps } from './validate-create-fs-props.function';

describe('validateCreateFSProps Function', () => {
  functionImportTest(validateCreateFSProps);
  toMatchSnapshot(() => createFsPropsSchema);

  it('should throw an error when passed value is not an object', () => {
    // @ts-ignore
    expect(() => validateCreateFSProps(null)).toThrowErrorMatchingSnapshot();
  });

  it('should require all object fields', () => {
    // @ts-ignore
    expect(() => validateCreateFSProps({})).toThrowErrorMatchingSnapshot();
  });

  it('should contain proper validators', () => {
    const props = {
      databaseName: 'd',
      databaseVersion: 1.1,
      rootDirectoryName: ' start_with_string',
      objectStoreName: '_contain_string_contain_string_contain_string_contain_string',
    };

    expect(() => validateCreateFSProps(props)).toThrowErrorMatchingSnapshot();
  });
});
