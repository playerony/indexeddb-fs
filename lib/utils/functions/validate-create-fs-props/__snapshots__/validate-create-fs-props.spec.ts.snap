// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateCreateFSProps Function should contain proper validators 1`] = `
"Props passed to createFS function are invalid:
[{\\"path\\":[\\"databaseName\\"],\\"property\\":\\"instance.databaseName\\",\\"message\\":\\"does not meet minimum length of 4\\",\\"schema\\":{\\"minLength\\":4,\\"maxLength\\":50,\\"type\\":\\"string\\",\\"pattern\\":\\"^[a-zA-Z0-9_.-]*$\\"},\\"instance\\":\\"d\\",\\"name\\":\\"minLength\\",\\"argument\\":4,\\"stack\\":\\"instance.databaseName does not meet minimum length of 4\\"},{\\"path\\":[\\"objectStoreName\\"],\\"property\\":\\"instance.objectStoreName\\",\\"message\\":\\"does not meet maximum length of 20\\",\\"schema\\":{\\"minLength\\":1,\\"maxLength\\":20,\\"type\\":\\"string\\",\\"pattern\\":\\"^[a-zA-Z0-9_.-]*$\\"},\\"instance\\":\\"_contain_string_contain_string_contain_string_contain_string\\",\\"name\\":\\"maxLength\\",\\"argument\\":20,\\"stack\\":\\"instance.objectStoreName does not meet maximum length of 20\\"},{\\"path\\":[\\"rootDirectoryName\\"],\\"property\\":\\"instance.rootDirectoryName\\",\\"message\\":\\"does not match pattern \\\\\\"^[a-zA-Z0-9_.-]*$\\\\\\"\\",\\"schema\\":{\\"minLength\\":1,\\"maxLength\\":20,\\"type\\":\\"string\\",\\"pattern\\":\\"^[a-zA-Z0-9_.-]*$\\"},\\"instance\\":\\" start_with_string\\",\\"name\\":\\"pattern\\",\\"argument\\":\\"^[a-zA-Z0-9_.-]*$\\",\\"stack\\":\\"instance.rootDirectoryName does not match pattern \\\\\\"^[a-zA-Z0-9_.-]*$\\\\\\"\\"},{\\"path\\":[\\"databaseVersion\\"],\\"property\\":\\"instance.databaseVersion\\",\\"message\\":\\"is not of a type(s) integer\\",\\"schema\\":{\\"minimum\\":1,\\"maximum\\":100,\\"type\\":\\"integer\\"},\\"instance\\":1.1,\\"name\\":\\"type\\",\\"argument\\":[\\"integer\\"],\\"stack\\":\\"instance.databaseVersion is not of a type(s) integer\\"}]"
`;

exports[`validateCreateFSProps Function should require all object fields 1`] = `
"Props passed to createFS function are invalid:
[{\\"path\\":[],\\"property\\":\\"instance\\",\\"message\\":\\"requires property \\\\\\"databaseName\\\\\\"\\",\\"schema\\":\\"/CreateFsPropsSchema\\",\\"instance\\":{},\\"name\\":\\"required\\",\\"argument\\":\\"databaseName\\",\\"stack\\":\\"instance requires property \\\\\\"databaseName\\\\\\"\\"},{\\"path\\":[],\\"property\\":\\"instance\\",\\"message\\":\\"requires property \\\\\\"databaseVersion\\\\\\"\\",\\"schema\\":\\"/CreateFsPropsSchema\\",\\"instance\\":{},\\"name\\":\\"required\\",\\"argument\\":\\"databaseVersion\\",\\"stack\\":\\"instance requires property \\\\\\"databaseVersion\\\\\\"\\"},{\\"path\\":[],\\"property\\":\\"instance\\",\\"message\\":\\"requires property \\\\\\"objectStoreName\\\\\\"\\",\\"schema\\":\\"/CreateFsPropsSchema\\",\\"instance\\":{},\\"name\\":\\"required\\",\\"argument\\":\\"objectStoreName\\",\\"stack\\":\\"instance requires property \\\\\\"objectStoreName\\\\\\"\\"},{\\"path\\":[],\\"property\\":\\"instance\\",\\"message\\":\\"requires property \\\\\\"rootDirectoryName\\\\\\"\\",\\"schema\\":\\"/CreateFsPropsSchema\\",\\"instance\\":{},\\"name\\":\\"required\\",\\"argument\\":\\"rootDirectoryName\\",\\"stack\\":\\"instance requires property \\\\\\"rootDirectoryName\\\\\\"\\"}]"
`;

exports[`validateCreateFSProps Function should throw an error when passed value is not an object 1`] = `
"Props passed to createFS function are invalid:
[{\\"path\\":[],\\"property\\":\\"instance\\",\\"message\\":\\"is not of a type(s) object\\",\\"schema\\":\\"/CreateFsPropsSchema\\",\\"instance\\":null,\\"name\\":\\"type\\",\\"argument\\":[\\"object\\"],\\"stack\\":\\"instance is not of a type(s) object\\"}]"
`;

exports[`validateCreateFSProps Function snapshot test should match snapshot 1`] = `
Object {
  "id": "/CreateFsPropsSchema",
  "properties": Object {
    "databaseName": Object {
      "maxLength": 50,
      "minLength": 4,
      "pattern": "^[a-zA-Z0-9_.-]*$",
      "type": "string",
    },
    "databaseVersion": Object {
      "maximum": 100,
      "minimum": 1,
      "type": "integer",
    },
    "objectStoreName": Object {
      "maxLength": 20,
      "minLength": 1,
      "pattern": "^[a-zA-Z0-9_.-]*$",
      "type": "string",
    },
    "rootDirectoryName": Object {
      "maxLength": 20,
      "minLength": 1,
      "pattern": "^[a-zA-Z0-9_.-]*$",
      "type": "string",
    },
  },
  "required": Array [
    "databaseName",
    "databaseVersion",
    "objectStoreName",
    "rootDirectoryName",
  ],
  "type": "object",
}
`;
